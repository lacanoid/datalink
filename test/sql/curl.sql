create table cu ( name text, url text );
insert into cu (url)
values
  ('file:///tmp/file1.txt'),
  ('file:/tmp/file1.txt'),
  ('file://localhost/tmp/file1.txt'),
  ('http://www.ljudmila.org/robots.txt'),
  ('http://www.ljudmila.org/robots2.txt'),
  ('data:,Hello%2C%20World%21'),
  ('data:text/plain;base64,SGVsbG8sIFdvcmxkIQ=='),
  ('data:text/html,%3Ch1%3EHello%2C%20World%21%3C%2Fh1%3E'),
  ('data:text/html,%3Cscript%3Ealert%28%27hi%27%29%3B%3C%2Fscript%3E'),
  ('file:/var/www/datalink/installcheck/utf8.txt'),
  ('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt')
;
-- select ok,rc,url,size,content_type,error from ( select (datalink.curl_perform(null,url)).* from cu ) as c;
-- test curl
select body,size,content_type from datalink.curl_perform(null,'data:,Hello%2C%20World%21');
select body,size,content_type from datalink.curl_perform(null,'data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==');
select body,size,content_type from datalink.curl_perform(null,'data:text/html,%3Ch1%3EHello%2C%20World%21%3C%2Fh1%3E');
select body,size,content_type from datalink.curl_perform(null,'data:text/html,%3Cscript%3Ealert%28%27hi%27%29%3B%3C%2Fscript%3E');

create server zala foreign data wrapper postgres_fdw options (dbname 'contrib_regression');
create user mapping for current_user server zala;
select url,ok,rc,error from datalink.curl_get('file://zala/etc/issue');
select url,ok,rc,error from datalink.curl_perform(null,'file://zala/etc/issue');
select body,ok,size,content_type,error from datalink.curl_perform(null,'file://zala/var/www/datalink/installcheck/utf8.txt');
drop user mapping for current_user server zala;
drop server zala;

select body,ok,size,content_type,error from datalink.curl_perform(null,'file:/var/www/datalink/installcheck/utf8.txt');

select body,ok,size,content_type,error from datalink.curl_perform(null,'https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt');

