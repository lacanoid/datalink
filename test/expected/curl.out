create table cu ( name text, url text );
insert into cu (url)
values
  ('file:///tmp/file1.txt'),
  ('file:/tmp/file1.txt'),
  ('file://localhost/tmp/file1.txt'),
  ('http://www.ljudmila.org/robots.txt'),
  ('http://www.ljudmila.org/robots2.txt'),
  ('data:,Hello%2C%20World%21'),
  ('data:text/plain;base64,SGVsbG8sIFdvcmxkIQ=='),
  ('data:text/html,%3Ch1%3EHello%2C%20World%21%3C%2Fh1%3E'),
  ('data:text/html,%3Cscript%3Ealert%28%27hi%27%29%3B%3C%2Fscript%3E'),
  ('file:/var/www/datalink/installcheck/utf8.txt'),
  ('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt')
;
-- select ok,rc,url,size,content_type,error from ( select (datalink.curl_perform(null,url)).* from cu ) as c;
-- test curl
select body,size,content_type from datalink.curl_perform(null,'data:,Hello%2C%20World%21');
     body      | size |        content_type         
---------------+------+-----------------------------
 Hello, World! |   13 | text/plain;charset=US-ASCII
(1 row)

select body,size,content_type from datalink.curl_perform(null,'data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==');
     body      | size | content_type 
---------------+------+--------------
 Hello, World! |   13 | text/plain
(1 row)

select body,size,content_type from datalink.curl_perform(null,'data:text/html,%3Ch1%3EHello%2C%20World%21%3C%2Fh1%3E');
          body          | size | content_type 
------------------------+------+--------------
 <h1>Hello, World!</h1> |   22 | text/html
(1 row)

select body,size,content_type from datalink.curl_perform(null,'data:text/html,%3Cscript%3Ealert%28%27hi%27%29%3B%3C%2Fscript%3E');
             body              | size | content_type 
-------------------------------+------+--------------
 <script>alert('hi');</script> |   29 | text/html
(1 row)

-- test some data: urls
select body,size,content_type from datalink.curl_get('data:,Hello%2C%20World%21');
     body      | size |        content_type         
---------------+------+-----------------------------
 Hello, World! |   13 | text/plain;charset=US-ASCII
(1 row)

select body,size,content_type from datalink.curl_get('data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==');
     body      | size | content_type 
---------------+------+--------------
 Hello, World! |   13 | text/plain
(1 row)

select body,size,content_type from datalink.curl_get('data:text/html,%3Ch1%3EHello%2C%20World%21%3C%2Fh1%3E');
          body          | size | content_type 
------------------------+------+--------------
 <h1>Hello, World!</h1> |   22 | text/html
(1 row)

select body,size,content_type from datalink.curl_get('data:text/html,%3Cscript%3Ealert%28%27hi%27%29%3B%3C%2Fscript%3E');
             body              | size | content_type 
-------------------------------+------+--------------
 <script>alert('hi');</script> |   29 | text/html
(1 row)

create server zala foreign data wrapper postgres_fdw options (dbname 'contrib_regression');
create user mapping for current_user server zala;
select url,ok,rc,error from datalink.curl_get('file://zala/etc/issue');
          url          | ok | rc | error 
-----------------------+----+----+-------
 file://zala/etc/issue | t  |  0 | 
(1 row)

select url,ok,rc,error from datalink.curl_perform(null,'file://zala/etc/issue');
          url          | ok | rc | error 
-----------------------+----+----+-------
 file://zala/etc/issue | t  |  0 | 
(1 row)

select body,ok,size,content_type,error from datalink.curl_perform(null,'file://zala/var/www/datalink/installcheck/utf8.txt');
             body              | ok | size | content_type | error 
-------------------------------+----+------+--------------+-------
 Math: ℕ ⊆ ℤ ⊆ ℚ ⊆ ℝ ⊆ ℂ ⊆ ℍ  +| t  |  206 |              | 
 Some 16bit emojis: ⛔⛳      +|    |      |              | 
 Some 32bit emojis: 😄😆😎🪙🩲+|    |      |              | 
 Some chinese: 𠀀𠀁𠀂𠀃       +|    |      |              | 
 〘_͡͡_͝͝_▹▸s͟σ͟͡ɴ͝ɪ͟c◂◃_͝͝_͡͡_〙          +|    |      |              | 
                               |    |      |              | 
(1 row)

drop user mapping for current_user server zala;
drop server zala;
select body,ok,size,content_type,error from datalink.curl_perform(null,'file:/var/www/datalink/installcheck/utf8.txt');
             body              | ok | size | content_type | error 
-------------------------------+----+------+--------------+-------
 Math: ℕ ⊆ ℤ ⊆ ℚ ⊆ ℝ ⊆ ℂ ⊆ ℍ  +| t  |  206 |              | 
 Some 16bit emojis: ⛔⛳      +|    |      |              | 
 Some 32bit emojis: 😄😆😎🪙🩲+|    |      |              | 
 Some chinese: 𠀀𠀁𠀂𠀃       +|    |      |              | 
 〘_͡͡_͝͝_▹▸s͟σ͟͡ɴ͝ɪ͟c◂◃_͝͝_͡͡_〙          +|    |      |              | 
                               |    |      |              | 
(1 row)

select body,ok,size,content_type,error from datalink.curl_perform(null,'https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt');
             body              | ok | size |       content_type        | error 
-------------------------------+----+------+---------------------------+-------
 Math: ℕ ⊆ ℤ ⊆ ℚ ⊆ ℝ ⊆ ℂ ⊆ ℍ  +| t  |  206 | text/plain; charset=utf-8 | 
 Some 16bit emojis: ⛔⛳      +|    |      |                           | 
 Some 32bit emojis: 😄😆😎🪙🩲+|    |      |                           | 
 Some chinese: 𠀀𠀁𠀂𠀃       +|    |      |                           | 
 〘_͡͡_͝͝_▹▸s͟σ͟͡ɴ͝ɪ͟c◂◃_͝͝_͡͡_〙          +|    |      |                           | 
                               |    |      |                           | 
(1 row)

